WebDevStudios/wd_s
# Adds pa11y-ci

### DESCRIPTION
This adds [pa11y-ci](https://github.com/pa11y/pa11y-ci) with some scripts to run pa11y on various environments to test for accessibility.

pa11y-ci uses pa11y to run its tests; by default it runs WCAG2AA which is actually WCAG 2.1 AA [according to the comment in this pull request](https://github.com/pa11y/pa11y/pull/603#issuecomment-900184208).

With this update, we're adding a new set of config variables to our package file:
```json
"config": {
	"local": "https://wdunderscores.test/",
	"staging": "https://wdunderscores.test/",
	"prod": "https://wdunderscores.test/"
},
```

Obviously, these would be updated for each project but I wanted to have them set in place at the onset so it's an easy path to replacement.

This also includes adding new scripts to utilize those config variables:
```json
"a11y:local": "pa11y-ci --sitemap $npm_package_config_local/sitemap.xml",
"a11y:prod": "pa11y-ci --sitemap $npm_package_config_prod/sitemap.xml",
"a11y:staging": "pa11y-ci --sitemap $npm_package_config_staging/sitemap.xml",
```

pa11y itself runs a test on a single page. pa11y-ci adds the ability to run a test on a sitemap. As WordPress should be creating a sitemap by default, we can assume that each of the environment URLs will have a sitemap generated. You shouldn't have to change these scripts at all since they pull from the variables, but you may need to add more config variables and more scripts if you have additional environments.

This also includes an update to our `sync` function to replace the hardcoded URL with a variable so you don't need to worry about replacing URLs in multiple places:
```json
"sync": "browser-sync start --https --proxy $npm_package_config_local --no-open --files 'build/*.*, **/*.html, **/*.php, !node_modules/**/*.html'",
```

### OTHER


### STEPS TO VERIFY
1. Check out the branch
2. `npm i --legacy-peer-deps`
3. Run `npm run a11y:local`
4. Feel free to change the config variables to other live sites if you want to test that out and make sure that the proper URLs are being used for the different scripts.

Right now, the results are just dumping into the command line.

There are additional ways to output results, like as JSON or HTML. We could discuss whether or not we'd want to add that for this now, or see if it's something we'd want to add down the road if it would add value. I haven't been able to really figure out how to get this to work for HTML, but I can dump JSON results to the console. Maybe we could use this as a way to parse JSON and automate the creation of Jira tickets for projects?

### DOCUMENTATION
Will this pull request require updating the wd_s [wiki](https://github.com/WebDevStudios/wd_s/wiki)?

For sure! If approved, we'll add a new page to the wiki to outline the setup and use of this.
